File: coding-assessments/leetcode/4sum-ii.py
from collections import Counter
def fourSumCount(A:[int], B:[int], C:[int], D:[int]) -> int:
    ab = Counter([a+b for a in A for b in B])
    cd = [c+d for c in C for d in D]
    count = 0
    for num in cd:
        count += ab[-num]
    return count;



fourSumCount(
A = [ 1, 2],
B = [-2,-1],
C = [-1, 2],
D = [ 0, 2]
)

File: coding-assessments/leetcode/4sum-ii.js

var fourSumCount = function (A, B, C, D) {
  let counter = new Map()
  for (a of A) {
    for (b of B) {
      const sum = a+b;
      counter.set(sum, counter.get(sum) + 1 || 1)
    }
  }

  let count = 0
  for (c of C) {
    for (d of D) {
      const sum = c+d;
      count += counter.get(-sum) || 0
    }
  }
  return count;
};
